components/edit.js

```diff
@@ -0,0 +1,235 @@
+    if (!isEdit && edit) {
+      setEditing( false )
+      router.replace(router.asPath)
+    }
+  }, [edit, router])
+
+  const editpage = <EditPage item={item}/>
+  const editlink = <EditLink item={item}/>
+
+  const componentToShow = edit ? editpage :
+  <>
+    {viewComponent}
+    {editlink}
+  </>
+  return componentToShow
+}
+
+export const EditableItem = ({ item, children }) => {
+  return useEditor(item, children)
+}
+
+EditableItem.propTypes = {
+  item: PropTypes.object.isRequired,
+}
+
+const SaveCancelControl = ({
+  cancelClick,
+  saveClick,
+  itemPath,
+  saveEnabled,
+}) => {
+  return (
+    <div className="flex space-x-1">
+      <Button.Link href={itemPath} navigationClick={cancelClick}>
+        Cancel
+      </Button.Link>
+      <Button
+        onClick={saveClick}
+        enabled={saveEnabled}
+        type={saveEnabled ? "primary" : "info"}
+      >
+        Save
+      </Button>
+    </div>
+  )
+}
+
+SaveCancelControl.propTypes = {
+  cancelClick: PropTypes.func.isRequired,
+  saveClick: PropTypes.func.isRequired,
+  itemPath: PropTypes.string.isRequired,
+  saveEnabled: PropTypes.bool.isRequired,
+}
+
+function sortedJson(obj) {
+  if (Array.isArray(obj)) {
+    return obj.map((value) => sortedJson(value))
+  }
+  // We know it's not an array if we're here because the above `if`
+  if (typeof obj == "object") {
+    const sorted = {}
+    Object.keys(obj)
+      .sort()
+      .forEach((key) => {
+        sorted[key] = obj[key]
+      })
+    return sorted
+  }
+  return obj
+}
+
+const SavedErrors = ({ errors = [] }) => {
+  return (
+    <div>
+      <p className="text-lg text-rose-600">Errors from Save:</p>
+      <ul className="list-disc">
+        {errors.map((error) => (
+          <li key={error.description} className="text-base text-rose-600">
+            {error.description}
+          </li>
+        ))}
+      </ul>
+    </div>
+  )
+}
+
+SavedErrors.propTypes = {
+  errors: PropTypes.array.isRequired,
+}
+
+const EditPage = ({ item }) => {
+  const path = item["@id"]
+
+  const loggedIn = useAuthenticated()
+
+  const editable = (session, item) => {
+    // cannot edit if not logged in or object not editable
+    const editable = canEdit(item)
+    return loggedIn && editable
+  }
+
+  const jsonErrors = (json) => {
+    // cannot save if we cannot edit or if the JSON is wrong
+    try {
+      JSON.parse(json)
+      return []
+    } catch (err) {
+      // Save the error
+      return [err.message]
+    }
+  }
+
+  const { session } = useContext(SessionContext)
+
+  const [text, setText] = useState(() => {
+    JSON.stringify({}, null, 4)
+  })
+
+  const [saveErrors, setErrors] = useState([])
+
+  const [editorStatus, setEditorStatus] = useState({
+    canEdit: editable(session, item),
+    canSave: editable(session, item),
+    errors: [],
+  })
+
+  useEffect(() => {
+    const fetch = new Fetch(session)
+    fetch.getObject(`${path}?frame=edit`, "GET").then((value) => {
```

Comment:
I think it’s confusing that you have to pass the "GET" operation to a method called `getObject()`. Would you ever pass "PUT" to `getObject()`?

Reply:
The intention here was that you could use any "safe" HTTP methods here (GET, HEAD, OPTIONS, and TRACE): from https://restfulapi.net/http-methods/.

But I guess there's strictly speaking nothing stopping you from saying "PUT".

If I was using a typed language, I'd maybe do something like:

```rust
enum HttpMethod {
    GET,
    HEAD,
    OPTIONS,
    TRACE,
    PUT,
    POST,
    PATCH
}

enum SafeMethod {
    GET,
    HEAD,
    OPTIONS,
    TRACE
}

impl From<SafeMethod> for HttpMethod {
    fn from(safe: SafeMethod) -> HttpMethod {
        match safe {
            SafeMethod::SAFE => HttpMethod::SAFE,
            SafeMethod::HEAD => HttpMethod::HEAD,
            SafeMethod::OPTIONS => HttpMethod::OPTIONS,
            SafeMethod::TRACE => HttpMethod::TRACE
        }
    }
}

fn getObject(path: Path, method: SafeMethod) -> Response {
    let http_method = HttpMethod::from(method);
    let headers = Headers::make_headers(). // ...etc
    fetch(path, http_method, headers)
}
```
In this way you'd guarantee that you could *only* use a safe http method with getObject. You type check on SafeMethod, which gets converted to a regular HttpMethod which gets sent to fetch. (I mean to be clear I'm making up this api in rust, just basing it off the javascript). But in JS you can just pass in anything, so I'm not the most aware what the best way to enforce something like this.

Does typescript do things like this?

Reply:
I believe Typescript lets you do that kind of type checking at build time. I decided against Typescript since we’re all new to NextJS and Tailwind CSS already, and didn’t want to add yet another thing that’s still somewhat controversial in the Javascript community.

Reply:
Maybe at least make the default value for the method argument "GET" since the other possible values are quite rare, at least for us so far.